Math/NumberTheory/Moduli/DiscreteLogarithm.hs

```diff
@@ -0,0 +1,104 @@
+discreteLogarithm' cg a b =
+  case cg of
+    CG2                                    -> 0
+       -- the only valid input was a=1, b=1
+    CG4                                    -> if b == 1 then 0 else 1
+       -- the only possible input here is a=3 with b = 1 or 3
+    CGOddPrimePower       (unPrime -> p) k -> discreteLogarithmPP p k (getVal a) (getVal b)
+    CGDoubleOddPrimePower (unPrime -> p) k -> discreteLogarithmPP p k (getVal a `rem` p^k) (getVal b `rem` p^k)
+       -- we have the isomorphism t -> t `rem` p^k from (Z/2p^kZ)* -> (Z/p^kZ)*
+
+-- Implementation of Bach reduction (https://www2.eecs.berkeley.edu/Pubs/TechRpts/1984/CSD-84-186.pdf)
+{-# INLINE discreteLogarithmPP #-}
+discreteLogarithmPP :: Integer -> Word -> Integer -> Integer -> Natural
+discreteLogarithmPP p 1 a b = discreteLogarithmPrime p a b
+discreteLogarithmPP p k a b = fromInteger result
+  where
+    baseSol    = toInteger $ discreteLogarithmPrime p (a `rem` p) (b `rem` p)
+    thetaA     = theta p pkMinusOne a
+    thetaB     = theta p pkMinusOne b
+    pkMinusOne = p^(k-1)
+    c          = (recipModInteger thetaA pkMinusOne * thetaB) `rem` pkMinusOne
+    result     = chineseRemainder2 (baseSol, p-1) (c, pkMinusOne)
+
+-- compute the homomorphism theta given in https://math.stackexchange.com/a/1864495/418148
+{-# INLINE theta #-}
+theta :: Integer -> Integer -> Integer -> Integer
+theta p pkMinusOne a = (numerator `quot` pk) `rem` pkMinusOne
+  where
+    pk           = pkMinusOne * p
+    p2kMinusOne  = pkMinusOne * pk
+    numerator    = (powModInteger a (pk - pkMinusOne) p2kMinusOne - 1) `rem` p2kMinusOne
+
+discreteLogarithmPrime :: Integer -> Integer -> Integer -> Natural
+discreteLogarithmPrime p a b
+  | p < 100000000 = fromIntegral $ discreteLogarithmPrimeBSGS (fromInteger p) (fromInteger a) (fromInteger b)
+  | otherwise     = discreteLogarithmPrimePollard p a b
+
+discreteLogarithmPrimeBSGS :: Int -> Int -> Int -> Int
+discreteLogarithmPrimeBSGS p a b = head [i*m + j | (v,i) <- zip giants [0..m-1], j <- maybeToList (M.lookup v table)]
+  where
+    m        = integerSquareRoot (p - 2) + 1 -- simple way of ceiling (sqrt (p-1))
+    babies   = iterate (.* a) 1
+    table    = M.fromList (zip babies [0..m-1])
+    aInv     = recipModInteger (toInteger a) (toInteger p)
+    bigGiant = fromInteger $ powModInteger aInv (toInteger m) (toInteger p)
+    giants   = iterate (.* bigGiant) b
+    x .* y   = x * y `rem` p
+
+discreteLogarithmPrimePollard :: Integer -> Integer -> Integer -> Natural
+discreteLogarithmPrimePollard p a b =
+  case concatMap runPollard [(0,0),(0,1),(1,1)] of
+    (t:_)  -> fromInteger t
+    []     -> error ("discreteLogarithm: pollard's rho failed, please report this as a bug. inputs " ++ show [p,a,b])
+  where
+    n                 = p-1 -- order of the cyclic group
+    halfN             = n `quot` 2
+    mul2 m            = if m < halfN then m * 2 else m * 2 - n
+    step (xi,!ai,!bi) = case xi `rem` 3 of
```

Comment:
Each `step` performs two divisions, which I expect to take a lot of time. I wonder whether we can avoid the first one. The algorithm is expected be robust to a somewhat unequal split into three subsets. 

* We can cast `xi` to `Word` and apply `rem` afterwards. At the very least we will avoid horribly expensive division of long `Integer` and in the best scenario compiler will apply [this trick](https://stackoverflow.com/questions/44212301/efficient-mod-3-in-x86-assembly).

* More aggressive approach is to skew distribution even further, looking at only 3 or 4 bits of `xi`. Something like this:

```haskell
case fromInteger xi .&. (7 :: Word) of 
  0 -> (xi*xi `rem` p, mul2 ai, mul2 bi)
  1 -> (xi*xi `rem` p, mul2 ai, mul2 bi)
  2 -> ( a*xi `rem` p,    ai+1,      bi)
  3 -> ( a*xi `rem` p,    ai+1,      bi)
  4 -> ( a*xi `rem` p,    ai+1,      bi)
  _ -> ( b*xi `rem` p,      ai,    bi+1)

Reply:
I like this idea! I recall reading that the expected time taken for Pollard can be improved to the expected random bound by using more partitions - I'm sure I read that tighter bounds were achieved with 20 but I can't seem to find that result now. With that in mind, it may be interesting to do something like `fromInteger xi .&. (15 :: Word)`, and have 16 different cases. 

In either case, would it be acceptable to take optimisation work like this into a different PR, and merge (something like) this one for now?

Reply:
Something like f_T and f_C from (Speeding up pollard's rho method for discrete logarithms, Teske 1998), or the walks described in 2.2.2 and 2.2.3 [here](https://maths-people.anu.edu.au/~brent/pd/rpb231.pdf).

Reply:
> In either case, would it be acceptable to take optimisation work like this into a different PR, and merge (something like) this one for now?

Sure. Could you please add a `-- TODO` comment with your ideas and links? It is easy to lose them in the depth of a closed PR.